### PR TITLE
Support git and stable salt-bootstrap on Gentoo.

### LINKED ISSUE #1499: "bootstrap-salt.sh git master" fails on Gentoo
### Description of Issue/Question
salt-bootstrap fails to install on Gentoo using the "git" method. Instead of bootstraping the code checked out from a GIT repository, it attempts to use `emerge` to install a non-existing ebuild `=app-admin/salt-9999`. This ebuild does not exist in the [Gentoo portage repository](https://gitweb.gentoo.org/repo/gentoo.git/tree/app-admin/salt).

### Setup
Gentoo system.

### Steps to Reproduce Issue
```
$ curl -o bootstrap-salt.sh -L https://bootstrap.saltstack.com
$ sudo sh bootstrap-salt.sh -D git master

 *  INFO: Running version: 2020.06.23
 *  INFO: Executed by: sh
 *  INFO: Command line: 'bootstrap-salt.sh -D git master'

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   5.4.38-gentoo
 *  INFO:   Distribution: Gentoo 2.7

 * DEBUG: Binaries will be searched using the following $PATH: /bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/opt/bin
 *  INFO: Installing minion

 *  WARN: Post Neon git based installations will always install salt
 *  WARN: and its dependencies using pip which will be upgraded to
 *  WARN: at least v8.0.0, and, in case the setuptools version is also
 *  WARN: too old, it will be upgraded to at least v9.1

 *  WARN: You have 10 seconds to cancel and stop the bootstrap process...

 * DEBUG: install_gentoo_2_git_deps not found....
 * DEBUG: install_gentoo_2_7_git_deps not found....
 * DEBUG: install_gentoo_2_deps not found....
 * DEBUG: install_gentoo_2_7_deps not found....
 *  INFO: Found function install_gentoo_git_deps
 * DEBUG: DEPS_INSTALL_FUNC=install_gentoo_git_deps
 * DEBUG: config_gentoo_2_git_salt not found....
 * DEBUG: config_gentoo_2_7_git_salt not found....
 * DEBUG: config_gentoo_2_salt not found....
 * DEBUG: config_gentoo_2_7_salt not found....
 * DEBUG: config_gentoo_git_salt not found....
 * DEBUG: config_gentoo_salt not found....
 *  INFO: Found function config_salt
 * DEBUG: CONFIG_SALT_FUNC=config_salt
 * DEBUG: preseed_gentoo_2_git_master not found....
 * DEBUG: preseed_gentoo_2_7_git_master not found....
 * DEBUG: preseed_gentoo_2_master not found....
 * DEBUG: preseed_gentoo_2_7_master not found....
 * DEBUG: preseed_gentoo_git_master not found....
 * DEBUG: preseed_gentoo_master not found....
 *  INFO: Found function preseed_master
 * DEBUG: PRESEED_MASTER_FUNC=preseed_master
 * DEBUG: install_gentoo_2_git not found....
 * DEBUG: install_gentoo_2_7_git not found....
 *  INFO: Found function install_gentoo_git
 * DEBUG: INSTALL_FUNC=install_gentoo_git
 * DEBUG: install_gentoo_2_git_post not found....
 * DEBUG: install_gentoo_2_7_git_post not found....
 * DEBUG: install_gentoo_2_post not found....
 * DEBUG: install_gentoo_2_7_post not found....
 * DEBUG: install_gentoo_git_post not found....
 *  INFO: Found function install_gentoo_post
 * DEBUG: POST_INSTALL_FUNC=install_gentoo_post
 * DEBUG: install_gentoo_2_git_restart_daemons not found....
 * DEBUG: install_gentoo_2_7_git_restart_daemons not found....
 * DEBUG: install_gentoo_2_restart_daemons not found....
 * DEBUG: install_gentoo_2_7_restart_daemons not found....
 * DEBUG: install_gentoo_git_restart_daemons not found....
 *  INFO: Found function install_gentoo_restart_daemons
 * DEBUG: STARTDAEMONS_INSTALL_FUNC=install_gentoo_restart_daemons
 * DEBUG: daemons_running_gentoo_2_git not found....
 * DEBUG: daemons_running_gentoo_2_7_git not found....
 * DEBUG: daemons_running_gentoo_2 not found....
 * DEBUG: daemons_running_gentoo_2_7 not found....
 * DEBUG: daemons_running_gentoo_git not found....
 * DEBUG: daemons_running_gentoo not found....
 *  INFO: Found function daemons_running
 * DEBUG: DAEMONS_RUNNING_FUNC=daemons_running
 * DEBUG: install_gentoo_2_git_check_services not found....
 * DEBUG: install_gentoo_2_7_git_check_services not found....
 * DEBUG: install_gentoo_2_check_services not found....
 * DEBUG: install_gentoo_2_7_check_services not found....
 * DEBUG: install_gentoo_git_check_services not found....
 *  INFO: Found function install_gentoo_check_services
 * DEBUG: CHECK_SERVICES_FUNC=install_gentoo_check_services
 *  INFO: Running install_gentoo_git_deps()
...
 *  INFO: Running install_gentoo_git()

These are the packages that would be merged, in order:

Calculating dependencies  
 * IMPORTANT: 7 news items need reading for repository 'gentoo'.
 * Use eselect news read to view new items.

... done!

emerge: there are no ebuilds to satisfy "=app-admin/salt-9999".
```

### Versions and Systems
bootstrap-salt.sh -- Version 2020.06.23
Gentoo system

### PR DESCRIPTION
**Pull request by ijansky**
### What does this PR do?
This PR fixes and improves Salt boostrap on Gentoo. The current implementation is outdated, does not support the git method, and does not have any tests on Gentoo.

### What issues does this PR fix or reference?
It fixes #1499.

### Previous Behavior
- It was not possible to bootstrap Salt using the "git" installation method. See #1499. 
- It was not possible to install a specific stable version - only one (unpredictable) version was supported.
- It was not possible to install the 2019.x version.
- No tests on Gentoo.

### New Behavior
- The "git" installation method is fully supported.
- It is possible to install any version available in Gentoo portage using the "stable" method - at the moment 2019.2, 3000.3, and 3001.1 versions.
- Salt correctly installs on Gentoo running either OpenRC or systemd.
- Added tests on Gentoo.

### CODE REVIEW

**Review by s0undt3ch — APPROVED** [review]

**Comment by s0undt3ch** [comment]
Coyld you please also add Gentoo [here](https://github.com/saltstack/salt-bootstrap/blob/develop/.github/workflows/templates/generate.py)?
That will add it to our CI process.

Thanks! By the way, awesome work!

**Comment by myii** [comment]
Previously, we tried to use the bootstrap to build pre-salted images for Gentoo but weren't successful.  I've since tried with this PR's branch and I was able to build images (based on [gentoo/stage3:systemd](https://hub.docker.com/layers/gentoo/stage3/systemd/images/sha256-4ebdf7dfb998b9757b4979c51c69532cafa6678932540da7f2ebdd6bc632837b?context=explore)) for:

* `git-master-py3`
  - Was able to complete the process for this and have it pushed by our `salt-image-builder` to Docker Hub (currently my fork):
    - https://hub.docker.com/r/myii/salt-master-py3/tags (tag `gentoo-stage3-systemd`)
* `stable-3001-py3`
* `stable-3000.3-py3`

Thanks for providing this, @ijansky!

**Comment by ijansky** [comment]
Thanks! I am glad to help.

@s0undt3ch PR was updated with the CI configuration. It adds quite a few tests to the pipeline. Is that ok or would a subset be more convenient?

@myii this is the next step. :) I will raise a PR with a Gentoo support to the https://gitlab.com/saltstack-formulas/infrastructure/salt-image-builder repository after this PR gets merged. You can check out the work in my fork https://gitlab.com/ijansky/salt-image-builder/-/tree/gentoo-support

**Comment by myii** [comment]
> @myii this is the next step. :) I will raise a PR with a Gentoo support to the https://gitlab.com/saltstack-formulas/infrastructure/salt-image-builder repository after this PR gets merged. You can check out the work in my fork https://gitlab.com/ijansky/salt-image-builder/-/tree/gentoo-support

@ijansky Nice work, it will be good to compare the two efforts.  Here's a rough outline of what I've used:

* https://gitlab.com/myii/salt-image-builder/-/compare/98e41413608c2e9ec581fb53bb669acb17c81899...b0318db31b3b0949bc9e2b7db09f118b8c897b98
  - Ignore the massive changes to `.gitlab-ci.yml`, since I commented out all of the other builds, so that I could focus on Gentoo.

Note that building `stable` has worked out _much_ slower and with more of our tests failing.  But we can pick up that conversation in the repo itself once the time comes.

File: bootstrap-salt.sh
Comment (ijansky) [inline bootstrap-salt.sh:7040]: Several tests failed due to the 20 minutes per build limit. Some time could be saved by using emerge-webrsync instead of emerge --sync, but that may not be sufficient.

Comment (s0undt3ch) [reply bootstrap-salt.sh:7040]: Should we add a `emerge-webrsync` to the kitchen config file for Gentoo?
At least its up to date when the test container is created....

File: .kitchen.yml
Comment (s0undt3ch) [inline .kitchen.yml:101]: Add `emerge-webrsync` here?

File: .kitchen.yml
Comment (s0undt3ch) [inline .kitchen.yml:107]: And here?

**Review by s0undt3ch — COMMENTED** [review]
This is what I was referring to

File: bootstrap-salt.sh
Comment (ijansky) [reply bootstrap-salt.sh:7040]: I like the idea of adding `emerge-webrsync` to the `provision_command` section, It is in the PR now.

Unfortunately it does not solve the problem. I went through the logs of one of the successful runs and it turned out that kitchen-docker runs `emerge --sync` when creating the container - see https://github.com/test-kitchen/kitchen-docker/blob/24a54be8dece9d095e0f9285c509def1c186f088/lib/kitchen/docker/helpers/dockerfile_helper.rb#L81-L88 

I guess that the solution is to replace `emerge --sync` by `emerge-webrsync` in kitchen-docker. I will play with it and open a PR against that code to see if it can be replaced or not.

EDIT: PR to kitchen-docker repository - https://github.com/test-kitchen/kitchen-docker/pull/373

Comment (ijansky) [reply bootstrap-salt.sh:7040]: With the changes in test-kitchen/kitchen-docker#373, more Gentoo builds succeed, but some of them still get cancelled. I guess that the only two options are either to skip running Gentoo tests in the pipeline or raise the 20 minutes build limit. What do you think, @s0undt3ch?

Comment (s0undt3ch) [reply bootstrap-salt.sh:7040]: We actually need to wait for the kitchen-docker release or install from git on the Gemfile

Comment (ijansky) [reply bootstrap-salt.sh:7040]: I updated the Gemfile to pull the latest kitchen-docker master in https://github.com/saltstack/salt-bootstrap/pull/1500/commits/821cc2a792a129f3a1ff14d46e6d127d8915344d. The last build was executed with these changes applied.

Comment (s0undt3ch) [reply bootstrap-salt.sh:7040]: Think 30 minutes will do?

Comment (myii) [reply bootstrap-salt.sh:7040]: @s0undt3ch These figures for building the pre-salted images from the bootstrap may be in the same sort of ballpark:

* 15 minutes 7 seconds: [`git-master-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759275102)
* 35 minutes 4 seconds: [`stable-3001-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759214114)
* 35 minutes 51 seconds: [`stable-3000.3-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759214115)

In fact, the respective equivalent to the first one took 13m 55s for [this PR](https://github.com/saltstack/salt-bootstrap/pull/1500/checks?check_run_id=1176492078).  I'd expect 35 minutes to be ample, especially after the `kitchen-docker` fix to avoid `emerge --sync`.

Comment (s0undt3ch) [reply bootstrap-salt.sh:7040]: Hmm, let's bump it to 40 minutes then :cry:

Comment (ijansky) [reply bootstrap-salt.sh:7040]: I am afraid that 45 minutes is closer to what these builds need.
- GIT and Salt >= 3000 ~ 17 minutes
- GIT and Salt 2019 ~ 40-45 minute
- Stable and Salt >= 3000 ~ 35-40 minutes
- Stable and Salt 2019 ~ 4045 minutes

We could drop 2019, since this one takes the most time - it needs to build Python 3.6.

Comment (s0undt3ch) [reply bootstrap-salt.sh:7040]: Can't drop 2019, it's still supported...

Comment (myii) [reply bootstrap-salt.sh:7040]: @s0undt3ch I was told by various SaltStack employees that the support is only until `Phase 3 support ends` on the following page:

* https://www.saltstack.com/product-support-lifecycle/

So it appears that `2019.2` is only supported for two more days.  Or is there something I'm misssing?

---

I suppose the same doesn't apply for the `salt-bootstrap`, which probably needs to continue until the `Extended life support ends`.

Comment (ijansky) [reply bootstrap-salt.sh:7040]: I made the job timeout configurable with the default set to 20 minutes unless overridden in a per distro map. Gentoo jobs have the timeout set to 45 minutes at the moment.

Good news is that all jobs succeeded and most Gentoo jobs finished in less than 20 minutes with just a few exceptions. When the support for 2019 gets dropped at some point in the future, the timeout could be decreased to around 30 minutes.

Let me know if I should squash all commits into one to keep the history cleaner.

**Comment by s0undt3ch** [comment]
Thank You!

**Pull request merged**
